test/uk/gov/hmrc/tai/connectors/HttpHandlerSpec.scala

```diff
@@ -46,286 +59,281 @@ class HttpHandlerSpec extends BaseSpec with BeforeAndAfterEach {
+    "should return a BadRequestException when BAD_REQUEST response" in {
+
+      server.stubFor(
+        put(anyUrl())
+          .willReturn(aResponse().withStatus(BAD_REQUEST)))
 
       val result = the[BadRequestException] thrownBy Await
-        .result(sut.putToApi[DateRequest]("", DateRequest(LocalDate.now())), 5.seconds)
+        .result(httpHandler.putToApi[DateRequest](testUrl, DateRequest(LocalDate.now())), 5.seconds)
 
       result.responseCode mustBe BAD_REQUEST
+
     }
 
-    "return Http exception" in {
-      when(http.PUT[DateRequest, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(GATEWAY_TIMEOUT)))
+    "should return a HttpException when unknown response" in {
+
+      server.stubFor(
+        put(anyUrl())
+          .willReturn(aResponse().withStatus(IM_A_TEAPOT).withBody("unknown response")))
+
+      val ex = the[Upstream4xxResponse] thrownBy Await
+        .result(httpHandler.putToApi[DateRequest](testUrl, DateRequest(LocalDate.now())), 5.seconds)
 
-      val result = the[HttpException] thrownBy Await
-        .result(sut.putToApi[DateRequest]("", DateRequest(LocalDate.now())), 5.seconds)
+      ex.message must include("unknown response")
 
-      result.responseCode mustBe GATEWAY_TIMEOUT
     }
   }
 
-  "postToApi" should {
+  "postToApi" must {
     val userInput = "userInput"
 
-    "return json which is coming from http post call" in {
+    List(
+      CREATED,
+      OK
+    ).foreach { httpStatus =>
+      s"return json which is coming from http post call for $httpStatus response" in {
 
-      when(http.POST[String, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-        .thenReturn(Future.successful(HttpResponse(OK, Some(Json.toJson(userInput)))))
-        .thenReturn(Future.successful(HttpResponse(CREATED, Some(Json.toJson(userInput)))))
+        server.stubFor(
+          post(anyUrl())
+            .willReturn(aResponse().withStatus(httpStatus).withBody(userInput)))
 
-      val okResponse = Await.result(sut.postToApi[String](mockUrl, userInput), 5 seconds)
-      val createdResponse = Await.result(sut.postToApi[String](mockUrl, userInput), 5 seconds)
+        val response = Await.result(httpHandler.postToApi[String](testUrl, userInput), 5 seconds)
 
-      okResponse.status mustBe OK
-      okResponse.json mustBe Json.toJson(userInput)
+        response.status mustBe httpStatus
 
-      createdResponse.status mustBe CREATED
-      createdResponse.json mustBe Json.toJson(userInput)
+      }
     }
 
-    "return Http exception" when {
-      "http response is NOT_FOUND" in {
-        when(http.POST[String, HttpResponse](any(), any(), any())(any(), any(), any(), any()))
-          .thenReturn(Future.successful(HttpResponse(NOT_FOUND)))
+    List(
+      GATEWAY_TIMEOUT,
+      INTERNAL_SERVER_ERROR,
+      SERVICE_UNAVAILABLE
+    ).foreach { httpStatus =>
+      s"return Http exception for $httpStatus response" in {
```

Comment:
This is not quite true, its returning a Upstream5xxResponse

Reply:
Changed name to Upstream5xxResponse